### PR TITLE
Colbert encoder types

### PR DESCRIPTION
**Pull request by avisil**
# Pull Request      

## What does this PR do?

```Closes #(issue)```       

Notes:
- Replace `(issue)` **above** ↑↑↑ with the issue this PR closes to automatically link the two.
This must be done when the PR is created.
- Add multiple `Closes #(issue)` as needed.
- If this PR is work towards but does not close an issue, simply tag the issue without mentioning `Closes`.

## Description

**Describe** the changes proposed by this PR below to give the reviewer context below ↓↓↓

```(description)```     

### Request Review

Be sure to **request a review** from one or more reviewers (unless the PR is to an unprotected branch).

## Versioning

When opening a PR to make changes to PrimeQA (i.e. `primeqa/`) master, be sure to increment the version following
[semantic versioning](https://semver.org/).  The VERSION is stored [here](https://github.com/primeqa/primeqa/blob/main/VERSION) 
and is incremented using `bump2version {patch,minor,major}` as described in the (development guide documentation)[development.html].

 - [ ] Have you updated the VERSION?
 - [ ] Or does this PR not change the `primeqa` package or was not into master?

After pulling in changes from master to an existing PR, ensure the VERSION is updated appropriately.
This may require bumping the version again if it has been previously bumped.

If you're not quite ready yet to post a PR for review, feel free to open a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

### After Merging
If merging into master and VERSION was updated, after this PR is merged:
- [ ] [Create a release](https://docs.github.com/en/github/administering-a-repository/releasing-projects-on-github/managing-releases-in-a-repository#creating-a-release) from the master with version equal to [VERSION](https://github.com/primeqa/primeqa/blob/main/VERSION)
- [ ] Not merging to master or VERSION not updated

### Checklist

Review the following and mark as completed:
- [ ] [Tag an issue or issues](#what-does-this-pr-do) this PR addresses.
- [ ] Added [description](#description) of changes proposed.
- [ ] Review requested as [appropriate](#request-review).
- [ ] Version bumped as [appropriate](#Versioning).
- [ ] New classes, methods, and functions documented.
- [ ] Documentation for modified code is updated.
- [ ] Built documentation to confirm it renders as expected (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Tests added to ensure all functionalities tested at >= 60% unit test coverage (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Release created [as needed](#after-merging) after merging.

### CODE REVIEW

File: primeqa/ir/dense/colbert_top/colbert/infra/config/settings.py
Comment (avisil) [inline primeqa/ir/dense/colbert_top/colbert/infra/config/settings.py:166]: Can we call them `encoder` family as opposed to `bert` ? Are we tied to BERT in general?

Comment (riyazbhat) [reply primeqa/ir/dense/colbert_top/colbert/infra/config/settings.py:166]: The code now supports models from three model families, `bert`, `roberta` and `xlm-roberta`.  Adding any encoder will be not feasible out of the box.

File: primeqa/components/indexer/dense.py
Comment (jdpsen) [inline primeqa/components/indexer/dense.py:46]: We should have a default set here - maybe "xlm-roberta"

File: primeqa/components/retriever/dense.py
Comment (jdpsen) [inline primeqa/components/retriever/dense.py:52]: Same here, We should have a default set here - maybe "xlm-roberta"

File: primeqa/ir/dense/colbert_top/colbert/infra/config/settings.py
Comment (jdpsen) [reply primeqa/ir/dense/colbert_top/colbert/infra/config/settings.py:166]: We should be consistent with the default setting what ever it is. I suggested xlm-roberta because PrimeQA leaderboard win is from Dr Decr which is xlm-roberta. Whatever, we choose that default is to be used everywhere.

**Review by jdpsen — COMMENTED** [review]
I tried with a checkpoint of model_type='roberta", both indexing and retrieval worked for me for a small dataset. Please make the default values consistent. 
Otherwise LGTM

File: primeqa/ir/dense/colbert_top/colbert/infra/config/settings.py
Comment (riyazbhat) [reply primeqa/ir/dense/colbert_top/colbert/infra/config/settings.py:166]: Makes sense.

File: primeqa/ir/dense/colbert_top/colbert/modeling/factory.py
Comment (franzmpub) [inline primeqa/ir/dense/colbert_top/colbert/modeling/factory.py:96]: Should this be `maxlen = colbert_config.teacher_doc_maxlen if is_teacher else colbert_config.doc_maxlen` ?

File: primeqa/ir/dense/colbert_top/colbert/training/training.py
Comment (franzmpub) [inline primeqa/ir/dense/colbert_top/colbert/training/training.py:125]: Is loading from existing LM still supported?

**Review by franzmpub — COMMENTED** [review]
It looks like the new code does not support backward compatibility with existing models represented as `.model` files?
Would it be enough to introduce mapping of the model type values (e.g. `roberta-base` -> `roberta`) ?

**Comment by riyazbhat** [comment]
> It looks like the new code does not support backward compatibility with existing models represented as `.model` files? Would it be enough to introduce mapping of the model type values (e.g. `roberta-base` -> `roberta`) ?

Its backward compatible with older checpoints, both .dnn and .model.

File: primeqa/ir/dense/colbert_top/colbert/training/training.py
Comment (riyazbhat) [reply primeqa/ir/dense/colbert_top/colbert/training/training.py:125]: What do you mean by existing LM? The code supports models on huggingface and models stored locally.

File: primeqa/ir/dense/colbert_top/colbert/modeling/factory.py
Comment (riyazbhat) [reply primeqa/ir/dense/colbert_top/colbert/modeling/factory.py:96]: Yes, I will fix this.

**Comment by franzmpub** [comment]
> 



> > It looks like the new code does not support backward compatibility with existing models represented as `.model` files? Would it be enough to introduce mapping of the model type values (e.g. `roberta-base` -> `roberta`) ?
> 
> Its backward compatible with older checpoints, both `.dnn` and `.model`.

Running `run_ir.py` with `--do_search` and an existing `roberta-base` model specified with `--model_name_or_path` results in 
`  File "primeqa/primeqa/ir/dense/colbert_top/colbert/modeling/factory.py", line 72, in get_colbert_from_pretrained
    raise NotImplementedError(f"Model type: {model_type} is not supported.")
`

**Review by franzmpub — COMMENTED** [review]
It looks like the usage of some argument/parameters has changed, e.g. using `--model_name_or_path` instead of `--init_from_lm`.
Is the plan to make the code backward compatible and/or update the related documentation.

File: primeqa/ir/dense/colbert_top/colbert/modeling/factory.py
Comment (jdpsen) [inline primeqa/ir/dense/colbert_top/colbert/modeling/factory.py]: @franzmpub : With this flag removed, wow are we supporting instantiating colBERT with a custom config now ?

Comment (franzmpub) [reply primeqa/ir/dense/colbert_top/colbert/modeling/factory.py]: @jdpsen (FYI @riyazbhat ): `model_type` will be the corresponding model type (e.g. `roberta` for WatBERT). 
The `custom` type was just inherited from the testing phase

Comment (franzmpub) [reply primeqa/ir/dense/colbert_top/colbert/modeling/factory.py]: @jdpsen (FYI @riyazbhat ): `model_type` will be the corresponding model type (e.g. `roberta` for WatBERT). 
The `custom` type was just inherited from the testing phase

**Review by franzmpub — APPROVED** [review]
LGTM

**Review by jdpsen — APPROVED** [review]
I'm good with the changes, good to merge once the CI tests work fine over the current main. 
Approved.

**Review by bhavani105 — COMMENTED** [review]
Martin would the services work without changes.  Is the model_type inferred given the path.

**Comment by franzmpub** [comment]
> Martin would the services work without changes. Is the model_type inferred given the path.

@riyazbhat any thoughts on that ^

**Comment by franzmpub** [comment]
> Martin would the services work without changes. Is the model_type inferred given the path.
 
@bhavani105 As far as I can tell, `model_type` related changes have been added to `primeqa/components/retriever/dense.py`, based on previous comments.

**Comment by riyazbhat** [comment]
> Martin would the services work without changes. Is the model_type inferred given the path.

No, model_type has to be explicitly passed as an argument.

**Comment by riyazbhat** [comment]
The custom model has been removed. So, LGTM.

**Review by riyazbhat — APPROVED** [review]

**Comment by bhavani105** [comment]
@riyazbhat How is the model_type  being set in the service layer.   Given a path to a ColBERT model, is it automatically set. Perhaps this should be tested before merging

File: primeqa/components/reranker/colbert_reranker.py
Comment (bhavani105) [inline primeqa/components/reranker/colbert_reranker.py]: @riyazbhat @franzmpub More on this 

This component as well as the ColBERTIndexer and ColBERTReranker are used in the service layer to load a ColBERT dnn/model file.  The service layer has no way of knowing what the  `model_type` so this cannot be required in the interface or set in the ColBERTConfig  in the `__post_init__` method.

Since the default is `xlm-roberta` it will work with the DrDecr model but it will fail on any model that is not xlm-roberta here https://github.com/primeqa/primeqa/blob/e25ece79a07f4c6fc85cb640a8d634b8acac8461/primeqa/ir/dense/colbert_top/colbert/modeling/factory.py#L45C5-L45C5

It is possible to look up the `model_type` in the dnn.  However i noticed for the drdecr model, this is set to `xlm-roberta-base`  not `xlm-roberta'   which will cause the code to fail with 
`NotImplementedError: Model type: xlm-roberta-base is not supported.` here https://github.com/primeqa/primeqa/blob/e25ece79a07f4c6fc85cb640a8d634b8acac8461/primeqa/ir/dense/colbert_top/colbert/modeling/factory.py#L72
`
It seems additional changes are need to fully support loading the models.  The components cannot require `model_type` to be specified.

Comment (franzmpub) [reply primeqa/components/reranker/colbert_reranker.py]: @riyazbhat (FYI @bhavani105) 
One possible solution is the map the model type as in the .dnn file to the corresponding model type as supported in the `if` statements starting here: [factory.py](https://github.com/primeqa/primeqa/blob/colbert_encoder_types/primeqa/ir/dense/colbert_top/colbert/modeling/factory.py#L53)  
This will require just a simple "many to one" mapping structure introduced in `factory.py` mapping e.g. "xlm-roberta-base", "xlm-roberta-large" and other "xlm-roberta-XYZ" flavors to "xlm-roberta".
What do you think?

Comment (bhavani105) [reply primeqa/components/reranker/colbert_reranker.py]: > @riyazbhat (FYI @bhavani105) One possible solution is the map the model type as in the .dnn file to the corresponding model type as supported in the `if` statements starting here: [factory.py](https://github.com/primeqa/primeqa/blob/colbert_encoder_types/primeqa/ir/dense/colbert_top/colbert/modeling/factory.py#L53) This will require just a simple "many to one" mapping structure introduced in `factory.py` mapping e.g. "xlm-roberta-base", "xlm-roberta-large" and other "xlm-roberta-XYZ" flavors to "xlm-roberta". What do you think?

Each model type has a corresponding HF config from which you can get a list of specific types e.g. 
https://github.com/huggingface/transformers/blob/b71f20a7c9f3716d30f6738501559acf863e2c5c/src/transformers/models/xlm_roberta/configuration_xlm_roberta.py#L27

Comment (franzmpub) [reply primeqa/components/reranker/colbert_reranker.py]: @riyazbhat , @bhavani105 : I've made changes in `factory.py` to allow mapping from the model subtypes to the types supported in `factory.py`.  Also made the default `model_type` in `components/reranker/colbert_reranker.py` set to `roberta`, same as in `components/retriever/dense.py`.
This now works for e.g. the DrDecr model, reading "xlm-roberta-base" from the model's config and mapping it to "xlm-roberta".

Comment (franzmpub) [reply primeqa/components/reranker/colbert_reranker.py]: @bhavani105 (FYI @riyazbhat ): I've made changes in `ir/dense/colbert_top/colbert/modeling/factory.py`, `components/reranker/colbert_reranker.py` and `components/retriever/dense.py` to avoid using the `model_type` parameter in the `components` classes.
Verified loading a model both from a checkpoint file and from a directory containing `config.json`.

File: primeqa/ir/README.md
Comment (bhavani105) [inline primeqa/ir/README.md:197]: Is this needed ?

File: primeqa/components/indexer/dense.py
Comment (bhavani105) [inline primeqa/components/indexer/dense.py]: Do we need model_type here ?

**Review by bhavani105 — COMMENTED** [review]
Remove model_type in ColBERTIndexer and in run_ir.py if possible

**Review by bhavani105 — APPROVED** [review]
LGTM

**Pull request merged**
